File: Practica3/QytetetJava/src/modeloqytetet/Casilla.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

/**
 *
 * @author Lorena
 */
public class Casilla {
    private int numeroCasilla; //indica numero de casilla
    private int coste; //indica el coste de compra del titulo de la propiedad de esa casilla
    private int numHoteles, numCasas; // indica el numero de hoteles y de casas edificados en esa casilla
    private TipoCasilla tipo; // indica el tipo de la casilla
    private TituloPropiedad titulo; //para asociar la casilla a su titulo de propiedad
    
    //Constructores:
   public Casilla() {
        
    }
    //Para los que son de tipo calle
    Casilla(int numeroCasilla, int coste, TipoCasilla tipo, TituloPropiedad titulo){
        this.numeroCasilla = numeroCasilla;
        this.coste = coste;
        numHoteles = 0;
        numCasas = 0;
        this.tipo = tipo;
        this.titulo = titulo;
    }
    
    //Para los que NO son de tipo calle
    Casilla(int numeroCasilla, TipoCasilla tipo){
        this.numeroCasilla = numeroCasilla;
        this.tipo = tipo;
    }


    
    //Modificadores
    void setNumHoteles(int nuevoNumero){
        numHoteles = nuevoNumero;
    }
    
    void setNumCasas(int nuevoNumero){
        numCasas = nuevoNumero;
    }
    
    private void setTitulo(TituloPropiedad titulo){
        this.titulo = titulo;
    }
    
    //Consultores
    int getNumeroCasilla(){
        return numeroCasilla;
    }
    
    int getCoste(){
        return coste;
    }
    
    int getNumHoteles(){
        return numHoteles;
    }
    
    int getNumCasas(){
        return numCasas;
    }
    
    public TipoCasilla getTipo(){
        return tipo;
    }
    
    public TituloPropiedad getTitulo(){
        return titulo;
    }
    
    int getCosteHipoteca(){
        return titulo.getHipotecaBase();
    }
    
    int getPrecioEdificar(){
        return titulo.getPrecioEdificar();
    }
    
    //Métodos
    TituloPropiedad asignarPropietario(Jugador jugador){
        titulo.setPropietario(jugador);
        return titulo;
    }
    
    int calcularValorHipoteca(){
        int hipotecaBase = titulo.getHipotecaBase();
        int cantidadRecibida = hipotecaBase + (int)(this.getNumCasas()*0.5*hipotecaBase + this.getNumHoteles()*hipotecaBase);
        
        return cantidadRecibida;
    }
    
    int cancelarHipoteca(){
        int cantidadRecibida = (int)(this.calcularValorHipoteca() + 0.1);
        return cantidadRecibida;
    }
    //Preguntar
    int cobrarAlquiler(){
        int costeAlquilerBase = titulo.getAlquilerBase();
        int costeAlquiler = costeAlquilerBase + (int)(this.getNumCasas()*0.5 + this.getNumHoteles()*2);
        
        titulo.cobrarAlquiler(costeAlquiler);
        
        return costeAlquiler; //?
    }
    
    int edificarCasa(){
        int nuevoNumero = this.getNumCasas() + 1;
        this.setNumCasas(nuevoNumero);
        
        int costeEdificarCasa = titulo.getPrecioEdificar();
        return costeEdificarCasa;
    }
    
    int edificarHotel(){
        setNumCasas(0);
        int nuevoNumero = this.getNumHoteles() + 1;
        this.setNumHoteles(nuevoNumero);
        
        int costeEdificarHotel = titulo.getPrecioEdificar();
        return costeEdificarHotel;
    }
    
    boolean estaHipotecada(){
        return titulo.getHipotecada();
    }
    
    int hipotecar(){
        titulo.setHipotecada(true);
        return this.calcularValorHipoteca();
    }
    
    int precioTotalComprar(){
        throw new UnsupportedOperationException("Sin implementar");
    }
    
    boolean propietarioEncarcelado(){
        return titulo.propietarioEncarcelado();
    }
    
    boolean sePuedeEdificarCasa(){
        return getNumCasas() < 4;
    }
    
    boolean sePuedeEdificarHotel(){
        if ( getNumCasas() == 4 && getNumHoteles() < 4){
            return true;
        }
        else{
            return false;
        }
    }
    
    boolean soyEdificable(){
        if (tipo == TipoCasilla.CALLE){
            return true;
        }
        else{
            return false;
        }
    }
    
    boolean tengoPropietario(){
        return titulo.tengoPropietario();
    }
    
    int venderTitulo(){
        int precioCompra = this.getCoste() + (this.getNumCasas() + this.getNumHoteles()) * titulo.getPrecioEdificar();
        int precioVenta = (int)(precioCompra + titulo.getFactorRevalorizacion()*precioCompra);
        
        titulo.setPropietario(null);
        this.setNumHoteles(0);
        this.setNumCasas(0);
        
        return precioVenta;
    }
    
    private void asignarTituloPropiedad(){}
    
    
    //Método toString() devuelve un String con el estado del objeto correspondiente
    @Override
    public String toString(){
        if(titulo != null){
            return "\n Numero de la Casilla = " + Integer.toString(numeroCasilla)
                    + "\n Coste = " + Integer.toString(coste)
                    + "\n Numero de hoteles = " + Integer.toString(numHoteles) 
                    + "\n Numero de casas = " + Integer.toString(numCasas)
                    + "\n Tipo de la casilla = " + tipo 
                    + "\n Titulo de propiedad = " + titulo.toString();
        }
        else{
            return "\n Numero de la Casilla = " + Integer.toString(numeroCasilla)
                + "\n Tipo de la casilla = " + tipo;

        }
    }

}

File: Practica1/Examen/QytetetRubyPruebaExamen/lib/casilla.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative 'tipo_casilla'
require_relative 'titulo_propiedad'

module ModeloQytetet
  class Casilla
    #Consultores y modificadores
    attr_accessor :num_hoteles, :num_casas, :titulo
    attr_reader :numero_casillas, :coste, :tipo
    
    #Constructores
    #Para los que son de tipo calle
    def initialize(numero_casilla, coste, tipo, titulo)
      @numero_casilla = numero_casilla
      @coste = coste
      @tipo = tipo
      @titulo = titulo
      @num_hoteles = 0
      @num_casas = 0
    end
    #Para los que NO son de tipo calle
    def self.no_calle(numero_casilla, tipo)
      Casilla.new(numero_casilla, 0, tipo, 0)
    end
    
    def get_coste_hipoteca
      return @titulo.hipotecaBase
    end
    
    def get_precio_edificar
      return @titulo.precioEdificar
    end
    
    #Métodos
    def asignar_propietario(jugador)
      return nil
    end
    
    def calcular_valor_hipoteca
      return nil
    end
    
    def cancelar_hipoteca
      return nil
    end
    
    def cobrar_alquiler
      return nil
    end
    
    def edificar_casa
      return nil
    end
    
    def edificar_hotel
      return nil
    end
    
    def esta_hipotecada
      return nil
    end
    
    def hipotecar
      return nil
    end
    
    def precio_total_comprar
      return nil
    end
    
    def propietario_encarcelado
      return nil
    end
    
    def se_puede_edificar_casa
      return nil
    end
    
    def se_puede_edificar_hotel
      return nil
    end
    
    def soy_edificable
      return nil
    end
    
    def tengo_propiedad
      return nil
    end
    
    def vender_titulo
      return nil
    end
    
    def asignar_titulo_propiedad
      return nil
    end
    #Método toString devuelve un String con el estado del objeto correspondiente
    def to_s
      if(@coste !=0)
        "\n Numero de la casilla: #{@numero_casilla} 
        \n Coste: #{@coste} 
        \n Numero de hoteles: #{@num_hoteles}
        \n Numero de casas: #{@num_casas}
        \n Tipo de la casilla: #{@tipo}
        \n Titulo de propiedad: #{@titulo}
        \n-----------------------------------------\n"
      else
      "\n Numero de la casilla: #{@numero_casilla}
      \n Tipo de la casilla: #{@tipo}
      \n-----------------------------------------\n"
      end
    end 
    private :titulo= 
    private :asignar_titulo_propiedad
  end
end
